pkg/chunk/cached_store.go

```diff
@@ -755,6 +770,88 @@ func parseObjOrigSize(key string) int {
 	return l
 }
 
+func (store *cachedStore) uploadStagingFile(key string, stagingPath string) int {
+	block, err := ioutil.ReadFile(stagingPath)
+	if err != nil {
+		logger.Errorf("open %s: %s", stagingPath, err)
+		return 0
+	}
+	buf := make([]byte, store.compressor.CompressBound(len(block)))
+	n, err := store.compressor.Compress(buf, block)
+	if err != nil {
+		logger.Errorf("compress chunk %s: %s", stagingPath, err)
+		return 0
+	}
+	compressed := buf[:n]
+
+	if strings.Count(key, "_") == 1 {
+		// add size at the end
+		key = fmt.Sprintf("%s_%d", key, len(block))
+	}
+	try := 0
+	for {
+		err := store.storage.Put(key, bytes.NewReader(compressed))
+		if err == nil {
+			break
+		}
+		logger.Infof("upload %s: %s (try %d)", key, err, try)
+		try++
+		time.Sleep(time.Second * time.Duration(try*try))
+	}
+	store.bcache.uploaded(key, len(block))
+	os.Remove(stagingPath)
+	return len(block)
+}
+
+func computeFreeGoal(innerStore *cacheStore, br float32) int64 {
+	total, _, _, _ := getDiskUsage(innerStore.dir)
+	var goal int64
+	toFree := int64(float32(total) * (innerStore.freeRatio - br))
+	if toFree > innerStore.used {
+		goal = innerStore.used
+	} else {
+		goal = toFree
+	}
+	return goal
+}
+
+func (store *cachedStore) cleanDelayStaging() {
+	cacheStoreKeys := make(map[*cacheStore][]string)
+
+	for key, _ := range store.pendingKeys {
```

Comment:
lock required here

Reply:
If the delete/insert on pendingKeys will not make the foreach fails , then no need to lock it.
We can check later